### PR TITLE
apps/examples/st_things : update example of message handling.

### PR DESCRIPTION
**Pull request by Hyunjun85**
This example can explain how to handle string array attribute from payload.

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [119c2cbf751e0d01de1844e3940a2ce0152e33aa] - Code Rule Check Result:
apps/examples/st_things/msghandling/sample_msghandling_user.c:240: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
apps/examples/st_things/msghandling/sample_msghandling_user.c:242: ERROR: [IDT_M_TAB] please, no spaces at the start of a line

**Comment by seinfra** [comment]
Target : [119c2cbf751e0d01de1844e3940a2ce0152e33aa] - Code Rule Check (C++) OK.

File: apps/examples/st_things/msghandling/contents/deivce_def.json [hunk 1]
```diff
@@ -200,4 +229,4 @@
       "privateKey": "privatekey"
     }
   }
-}
+}
```
Comment (juitem) [inline apps/examples/st_things/msghandling/contents/deivce_def.json]: add new line at the end of line

Comment (Hyunjun85) [reply apps/examples/st_things/msghandling/contents/deivce_def.json]: fixed.

**Comment by seinfra** [comment]
Target : [eebac18c48b6d2135e4442691bfd8711656bffa7] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [eebac18c48b6d2135e4442691bfd8711656bffa7] - Code Rule Check OK.

**Comment by YiHyunJin** [comment]
URI_TURBO_LEN is Intuitive so easy to see.

File: apps/examples/st_things/msghandling/sample_msghandling_user.c [hunk 2]
```diff
@@ -174,10 +202,42 @@ bool handle_set_request_on_resource_wwstfilterstateresuri(st_things_set_request_
...
+		resp_rep->set_str_array_value(resp_rep, PROP_MODES, (const char **)g_modes, g_mode_cnt);
+	}
+	if (req_msg->has_property_key(req_msg, PROP_SUPPORTEDMODES)) {
+		resp_rep->set_str_array_value(resp_rep, PROP_SUPPORTEDMODES, (const char **)g_supported_modes, MAX_SUPPORTED_MODES);
+	}
+
+	return true;
+}
+
+bool handle_set_request_on_resource_wwstrobotcleanerturbomoderesuri(st_things_set_request_message_s* req_msg, st_things_representation_s* resp_rep)
```
Comment (JeongJunSik) [inline apps/examples/st_things/msghandling/sample_msghandling_user.c]: Coding rules check (* should be attached to the variable side)

File: apps/examples/st_things/msghandling/sample_msghandling_user.c [hunk 3]
```diff
@@ -174,10 +202,42 @@ bool handle_set_request_on_resource_wwstfilterstateresuri(st_things_set_request_
...
-		strncpy(g_machine_state, machinestates, 10);
+		strncpy(g_machine_state, machinestates, MAX_STR_LEN);
 		resp_rep->set_str_value(resp_rep, PROP_MACHINESTATES, machinestates);
 		st_things_notify_observers(req_msg->resource_uri);
 	}
 
 	return true;
 }
+
+bool handle_get_request_on_resource_wwstrobotcleanerturbomoderesuri(st_things_get_request_message_s* req_msg, st_things_representation_s* resp_rep)
```
Comment (JeongJunSik) [inline apps/examples/st_things/msghandling/sample_msghandling_user.c]: Coding rules check (* should be attached to the variable side)

File: apps/examples/st_things/msghandling/sample_msghandling_things.c [hunk 4]
```diff
@@ -47,20 +47,31 @@
...
+
 extern bool handle_get_request_on_resource_wwstairqualitysensorlevelresuri(st_things_get_request_message_s *req_msg, st_things_representation_s *resp_rep);
+
 extern bool handle_get_request_on_resource_tone(st_things_get_request_message_s *req_msg, st_things_representation_s *resp_rep);
 extern bool handle_set_request_on_resource_tone(st_things_set_request_message_s *req_msg, st_things_representation_s *resp_rep);
 
 extern bool handle_get_request_on_resource_wwstfilterstateresuri(st_things_get_request_message_s *req_msg, st_things_representation_s *resp_rep);
 extern bool handle_set_request_on_resource_wwstfilterstateresuri(st_things_set_request_message_s *req_msg, st_things_representation_s *resp_rep);
 
+extern bool handle_get_request_on_resource_wwstrobotcleanerturbomoderesuri(st_things_get_request_message_s* req_msg, st_things_representation_s* resp_rep);
```
Comment (JeongJunSik) [inline apps/examples/st_things/msghandling/sample_msghandling_things.c]: Coding rules check (* should be attached to the variable side)

File: apps/examples/st_things/msghandling/sample_msghandling_things.c [hunk 5]
```diff
@@ -47,20 +47,31 @@
...
+extern bool handle_set_request_on_resource_wwstrobotcleanerturbomoderesuri(st_things_set_request_message_s* req_msg, st_things_representation_s* resp_rep);
```
Comment (JeongJunSik) [inline apps/examples/st_things/msghandling/sample_msghandling_things.c]: Coding rules check (* should be attached to the variable side)

**Comment by Hyunjun85** [comment]
@JeongJunSik  I have fixed some codes. please review again.

File: apps/examples/st_things/msghandling/sample_msghandling_things.c [hunk 6]
```diff
@@ -47,20 +47,31 @@
...
 extern bool handle_get_request_on_resource_tone(st_things_get_request_message_s *req_msg, st_things_representation_s *resp_rep);
 extern bool handle_set_request_on_resource_tone(st_things_set_request_message_s *req_msg, st_things_representation_s *resp_rep);
 
 extern bool handle_get_request_on_resource_wwstfilterstateresuri(st_things_get_request_message_s *req_msg, st_things_representation_s *resp_rep);
 extern bool handle_set_request_on_resource_wwstfilterstateresuri(st_things_set_request_message_s *req_msg, st_things_representation_s *resp_rep);
 
+extern bool handle_get_request_on_resource_wwstrobotcleanerturbomoderesuri(st_things_get_request_message_s *req_msg, st_things_representation_s *resp_rep);
+extern bool handle_set_request_on_resource_wwstrobotcleanerturbomoderesuri(st_things_set_request_message_s *req_msg, st_things_representation_s *resp_rep);
+
 #ifdef CONFIG_RESET_BUTTON
```
Comment (JeongJunSik) [inline apps/examples/st_things/msghandling/sample_msghandling_things.c]: if you want using this config, you need to set RESET_BUTTION depends on

File: apps/examples/st_things/msghandling/sample_msghandling_user.c [hunk 7]
```diff
@@ -174,10 +202,42 @@ bool handle_set_request_on_resource_wwstfilterstateresuri(st_things_set_request_
...
+	if (req_msg->has_property_key(req_msg, PROP_SUPPORTEDMODES)) {
+		resp_rep->set_str_array_value(resp_rep, PROP_SUPPORTEDMODES, (const char **)g_supported_modes, MAX_SUPPORTED_MODES);
+	}
+
+	return true;
+}
+
+bool handle_set_request_on_resource_wwstrobotcleanerturbomoderesuri(st_things_set_request_message_s *req_msg, st_things_representation_s *resp_rep)
+{
+
```
Comment (JeongJunSik) [inline apps/examples/st_things/msghandling/sample_msghandling_user.c]: remove this line

**Comment by Hyunjun85** [comment]
@JeongJunSik Some codes have changed. please review again.

**Comment by seinfra** [comment]
Target : [00bdce522ac203ffc968255d9862b7dc75019a17] -

**Comment by seinfra** [comment]
Target : [00bdce522ac203ffc968255d9862b7dc75019a17] -

File: apps/examples/st_things/msghandling/sample_msghandling_user.c [hunk 8]
```diff
@@ -52,6 +62,29 @@ static int64_t g_color_temp_range[MAX_STR_LEN]      = {0, 100};
...
+void init_modes(void)
+{
+	int idx;
+
+	g_mode_cnt = 1;
+	g_modes = (char **)malloc(sizeof(char*) * 1);
+	g_modes[0] = malloc(sizeof(char) * MAX_STR_LEN);
+	strncpy(g_modes[0], g_machine_state, MAX_STR_LEN);
+
+	g_supported_modes = (char **)malloc(sizeof(char*) * MAX_SUPPORTED_MODES);
```
Comment (sunghan-chang) [inline apps/examples/st_things/msghandling/sample_msghandling_user.c]: I can't find freeing "g_supported_modes" and "element".

File: apps/examples/st_things/msghandling/sample_msghandling_user.c [hunk 9]
```diff
@@ -174,10 +202,41 @@ bool handle_set_request_on_resource_wwstfilterstateresuri(st_things_set_request_
...
+
+bool handle_set_request_on_resource_wwstrobotcleanerturbomoderesuri(st_things_set_request_message_s *req_msg, st_things_representation_s *resp_rep)
+{
+	char **req_str_array;
+	size_t array_size;
+	int idx;
+
+	if (req_msg->rep->get_str_array_value(req_msg->rep, PROP_MODES, &req_str_array, &array_size)) {
+		for (idx = 0; idx < g_mode_cnt; ++idx) {
+			free(g_modes[idx]);
```
Comment (sunghan-chang) [inline apps/examples/st_things/msghandling/sample_msghandling_user.c]: I saw only two allocation for g_modes.
One is for g_modes, another is g_modes[0].
Why do you free g_modes with g_mode_cnt?

**Comment by seinfra** [comment]
Target : [d1e31a2cdadb6404ec565a29d246d6a659429dae] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [d1e31a2cdadb6404ec565a29d246d6a659429dae] - Code Rule Check OK.

File: apps/examples/st_things/msghandling/sample_msghandling_user.c (hunk 9 above)
Comment (Hyunjun85) [reply apps/examples/st_things/msghandling/sample_msghandling_user.c]: In the init_modes(), g_modes  is set char* array and got an element.
And then g_modes, g_mode_cnt are updated when it receivces post reqeust from the client.
So the setted datas must be freed.

File: apps/examples/st_things/msghandling/sample_msghandling_user.c (hunk 8 above)
Comment (Hyunjun85) [reply apps/examples/st_things/msghandling/sample_msghandling_user.c]: This is just a example to show how to handle request message.
g_supported_modes is living during the app running. 
If g_supported_modes is freed, the app will get a strange action.

**Pull request merged**